include/exadg/aero_acoustic/calculators/source_term_calculator.h

```diff
@@ -0,0 +1,137 @@
+#ifndef INCLUDE_EXADG_AERO_ACOUSTIC_SOURCE_TERM_CALCULATOR_H_
+#define INCLUDE_EXADG_AERO_ACOUSTIC_SOURCE_TERM_CALCULATOR_H_
+
+#include <exadg/matrix_free/integrators.h>
+#include <exadg/utilities/lazy_ptr.h>
+
+namespace ExaDG
+{
+namespace AeroAcoustic
+{
+struct SourceTermCalculatorData
+{
+  unsigned int dof_index_pressure;
+  unsigned int dof_index_velocity;
+  unsigned int quad_index;
+
+  // density of the underlying fluid.
+  double density;
+
+  // use material or partial temporal derivative of pressure as source term.
+  bool consider_convection;
+};
+
+/**
+ * A class that knows how to compute the aeroacoustic source term on the fluid mesh.
+ * evaluate_integrate() computes and integrates the source term on the fluid mesh.
+ *
+ * The aeroacoustic source term f is definded as:
+ * f = - rho * (dp/dt + u * grad(p)).
+ * The scaling factor rho has to be used since the pressure of the incompressible
+ * module is a kinematic pressure. Using consider_convection=false
+ * f = -rho * (dp/dt).
+ */
+template<int dim, typename Number>
+class SourceTermCalculator
+{
+  using This       = SourceTermCalculator<dim, Number>;
+  using VectorType = dealii::LinearAlgebra::distributed::Vector<Number>;
+
+  using CellIntegratorScalar = CellIntegrator<dim, 1, Number>;
+  using CellIntegratorVector = CellIntegrator<dim, dim, Number>;
+
+public:
+  SourceTermCalculator() : matrix_free(nullptr)
+  {
+  }
+
+  void
+  setup(dealii::MatrixFree<dim, Number> const & matrix_free_in,
+        SourceTermCalculatorData const &        data_in)
+  {
+    matrix_free = &matrix_free_in;
+    data        = data_in;
+  }
+
+  void
+  evaluate_integrate(VectorType &       dst,
+                     VectorType const & velocity_cfd_in,
+                     VectorType const & pressure_cfd_in,
+                     VectorType const & pressure_cfd_time_derivative_in)
+  {
+    dst.zero_out_ghost_values();
+
+    if(data.consider_convection)
+    {
+      velocity_cfd.reset(velocity_cfd_in);
+      velocity_cfd->update_ghost_values();
+
+      pressure_cfd.reset(pressure_cfd_in);
+      pressure_cfd->update_ghost_values();
+
+      matrix_free->cell_loop(&This::compute_integrated_source_term<true>,
+                             this,
+                             dst,
+                             pressure_cfd_time_derivative_in,
+                             true);
+    }
+    else
+    {
+      matrix_free->cell_loop(&This::compute_integrated_source_term<false>,
+                             this,
+                             dst,
+                             pressure_cfd_time_derivative_in,
+                             true);
+    }
+  }
+
+private:
+  template<bool consider_convection>
```

Comment:
Personally, I do not see the reasoning for such templates. Here we compute a coupling term, which is done rarely compared to "matrix-vector products" / operator evaluations done often in the single field solvers. So from a performance perspective, I see even less motivation here to make it a template parameter. A few lines above one can also see that using templates for this purpose make the code harder to read (a procedure, that can in my opinion not really be extended to cases where we have more than one such parameter).

Reply:
[fb6bc6d](https://github.com/exadg/exadg/pull/635/commits/fb6bc6d5665cee32521fb32710b671ebd67953b2)